src/plans/constrained_plan.jl

```diff
@@ -35,240 +49,102 @@ Describes the constrained objective
+* `equality_constraints`: an [`AbstractManifoldObjective`](@ref) representing the equality constraints
+``h: \mathcal M → \mathbb R^n`` also possibly containing its gradient and/or Hessian
 
 # Constructors
-
-    ConstrainedManifoldObjective(f, grad_f, g, grad_g, h, grad_h;
-        evaluation=AllocatingEvaluation()
+    ConstrainedManifoldObjective(M::AbstractManifold, f, grad_f;
+        g=nothing,
+        grad_g=nothing,
+        h=nothing,
+        grad_h=nothing;
+        evaluation=AllocatingEvaluation(),
     )
 
-Where `f, g, h` describe the cost, inequality and equality constraints, respectively, as
-described previously and `grad_f, grad_g, grad_h` are the corresponding gradient functions in
-one of the 4 formats. If the objective does not have inequality constraints, you can set `G` and `gradG` no `nothing`.
-If the problem does not have equality constraints, you can set `H` and `gradH` no `nothing` or leave them out.
+    ConstrainedManifoldObjective(M::AbstractManifold, mho::AbstractManifoldObjective;
+        equality_constraints::Union{Nothing, AbstractManifoldObjective} = nothing,
+        inequality_constraints::Union{Nothing, AbstractManifoldObjective} = nothing
+    ) where {IMO <: Union{Nothing, AbstractManifoldObjective} EMO <: Union{Nothing, AbstractManifoldObjective}}
 
-    ConstrainedManifoldObjective(M::AbstractManifold, F, gradF;
-        G=nothing, gradG=nothing, H=nothing, gradH=nothing;
-        evaluation=AllocatingEvaluation()
-    )
-
-A keyword argument variant of the preceding constructor, where you can leave out either
-`G` and `gradG` or `H` and `gradH` but not both pairs.
+TODO: Describe constructors
 """
 struct ConstrainedManifoldObjective{
-    T<:AbstractEvaluationType,CT<:ConstraintType,TCost,GF,TG,GG,TH,GH
-} <: AbstractManifoldGradientObjective{T,TCost,GF}
-    cost::TCost
-    gradient!!::GF
-    g::TG
-    grad_g!!::GG
-    h::TH
-    grad_h!!::GH
-end
-#
-# Constructors I: functions
-#
+    T<:AbstractEvaluationType,
+    CT<:AbstractConstraintType,
+    MO<:AbstractManifoldObjective,
+    IMO<:Union{Nothing,AbstractManifoldObjective},
+    EMO<:Union{Nothing,AbstractManifoldObjective},
+} <: AbstractManifoldObjective{T}
+    objective::MO
+    inequality_constrinats::IMO
+    equality_constraints::EMO
+end
+# Generic f, grad_f -> pass on to new constructor
 function ConstrainedManifoldObjective(
-    f::TF,
-    grad_f::TGF,
-    g::Function,
-    grad_g::Function,
-    h::Function,
-    grad_h::Function;
+    f,
+    grad_f,
+    g,
+    grad_g,
+    h,
+    grad_h;
     evaluation::AbstractEvaluationType=AllocatingEvaluation(),
-) where {TF,TGF}
-    return ConstrainedManifoldObjective{
-        typeof(evaluation),
-        FunctionConstraint,
-        TF,
-        TGF,
-        typeof(g),
-        typeof(grad_g),
-        typeof(h),
-        typeof(grad_h),
-    }(
-        f, grad_f, g, grad_g, h, grad_h
+    constraint::AbstractConstraintType=if all(
```

Comment:
I'm trying to fix this but I don't quite understand what this default value is supposed to do. Could you help a bit here?

Reply:
Ah no, sorry my bad (too many meetings today). Of course the other case is `ComponentVectorialType` because the function components are given if it is a vector of functions.

Reply:
Hm, that doesn't look great. IMO the primary internal representation of multiple constraints should be a function with constraint number as an additional indexing argument. It usually has lower performance overhead and if you really like a vector of functions, you can still have a wrapper for that. The only downside is that it's no longer possible to guess which type of constraint we have. What do you think about making such a rework?

Reply:
Uff, what you propose to wrap that again is slowly turning into a matriska then. I mean `g` should here be either a function _or a vector of functions.
You would propose it is a function or a struct containing a vector of functions? 

The advantage of a vector of functions is twofold: (1) yes one can directly read off the number of components there, but (2) one can evaluate them individually, which _really_ speeds up ALM and EPM, since both require efficient single constraint access (both function and gradient) instead of evaluation _all_ constraints to then access the `i`th one.

So I do not see a good reason to do a wrapper here.

Reply:
You can still evaluate single constraints in my approach but using `co.g(M, p, i)` instead of `co.g(M, p)[i]`. You could evaluate all of them using `co.g(M, p, :)` or an arbitrary subset if needed: `co.g(M, p, [1, 2, 10])`. Shared computations (between constraints) are easy to handle and there are no type instabilities due to vectors of functions of different types.

Reply:
Yes your two signatures are nearly the two forms we currently have `co.g(M,p)[j]` for the single function and `co.g[j](M,p)`for the vector...but I still do not get why you want to wrap `g` further, we would then

* have the constraintObjective which contains
* the new vector type here which then
* contains a new special type to wrap g and to wrap grad_g?

So instead of one struct (before this PR we only had the very first one) we then would have 3? Or do you wrap something more inside the to most inner structs?
I mean if even I feel we get lost in struct-ception ;)

Reply:
> but I still do not get why you want to wrap `g` further, we would then

Because I don't want to ever use vector-of-functions representation for constraints. I'm fine with any design that allows me to bypass it.

Reply:
But you can always provide `g(M,p)` and `grad_g(M,p)` (for now only mapping into nested power on main) and be happy that ALM/EPM is n-times slower than for the vector version (n the number of constraints).

Reply:
I completely don't see why ALM/EPM would be slower, you can still evaluate single constraints, just with a different interface.

Reply:
I mean propose your version with another struct – I just hope it stays a finite number of struct we wrap than and that access is something that is doable. But it has to also be compatible with the nested variant, otherwise this is breaking.

I currently am not able to achieve both with my design ideas. So if your (hopefully finitely many wrapped) structs achieve that, sure. We can do that.

Reply:
OK, I will try to come up with something.